### ISSUE TITLE
Stray encrypted files

### ISSUE THREAD

**Issue by loungebob** [bailey27/cppcryptfs#109]
I dont really know where to start or how this happens but maybe someone cleverer than me can figure this out.

I use win10, google drive and google drive file stream and cppcryptfs.

google drive is mounted under windows via drive file stream. folder structure “google drive file stream/my drive/folder1/folder2”

folder2 is mounted via cppcryptfs under windows and is the encrypted folder.

all file transactions are done via the cppcryptfs mounted “clear” folder. drive file stream syncs all.

this works mostly well except for the last few weeks, maybe even longer I find encrypted files and folders in the “my drive” (google drive root) folder.

I have no idea how they get there since only folder2 is mounted in cppcryptfs and I only work with the mounted folder2 in windows file explorer. 

since finding out what those encrypted folders and files are is very cumbersome I wondered how this happens and if it’s a cppcryptfs or google drive file stream bug and how to prevent this from happening.

also, is there a quick way to find out what those stray files and folders are, make them clear text? it’s a couple of hundred files, I dont want to do the whole manually rename them with a valid encrypted file name trick.

to summarize: I upload files to google drive via drive file stream into a cppcryptfs encrypted and in windows file explorer mounted folder. sometimes those seem to end up 2 folders up in the hierarchy (where cppcryptfs shouldnt even have access).

any ideas whats happening here?

**Comment by bailey27** [comment]
If the files exist in both the root of you google file stream drive and also exist somewhere under "google drive file stream/my drive/folder1/folder2", then you need to find them under "google drive file stream/my drive/folder1/folder2" to get the right gocryptf.diriv file so you can decrypt the file name.

Make a new local folder,e.g c:\tmp\recover

Copy the gocryptfs.conf and gocryptfs.diriv from "google drive file stream/my drive/folder1/folder2" to c:\tmp\recover

On the cppcryptfs settings page, change the cache time to live to 1 sec (so you won't have to wait for very long for the diriv to expire in in the cache).

Mount c:\tmp\recover to some drive letter.

Run widows cmd.exe

cd to google drive file stream/my drive/folder1/folder2
look try to find one of the stray files files under there 
e.g.

dir /s/b PNPm_jLaytQyYaz5ukn-yA

In my test case, it found the file in this path

C:\tmp\test\xN0sW_ZCI1fBcmTd7jxq-Q\PNPm_jLaytQyYaz5ukn-yA

copy the diriv file from that directory to c:\tmp\recover (over the gocryptfs that was there before).
e.g.
```
copy c:\tmp\test\xN0sW_ZCI1fBcmTd7jxq-Q\gocryptfs.diriv c:\tmp\recover
Overwrite c:\tmp\recover\gocryptfs.diriv? (Yes/No/All): y
        1 file(s) copied.
```
copy the file to c:\tmp\recover

```
copy C:\tmp\test\xN0sW_ZCI1fBcmTd7jxq-Q\PNPm_jLaytQyYaz5ukn-yA c:\tmp\recover
```

Now, if everything went well, if you look on the mounted drive letter for c:\tmp\recover you should see the file with the uncencrypted name, and it should be accessible.

You could also decrypt all the files at once (but not the names) fairly easily.

Here is the procedure.

Create an empty folder on your local hard drive. e.g. c:\tmp\recover2

Copy the gocryptfs.conf from your encrypted filesystem to c:\tmp\recover2

Edit the gocryptfs.conf in notepad and make these changes.

Change the line that says 
```
"EMENames",
``` 
to say 
```
"PlaintextNames",
```
Be careful to have the quotes and comma there after your edit.

Delete the lines that say

```
"LongNames",
"DirIV",
```

Save the gocryptfs.conf.

Now, copy the stray encrypted files into c:\tmp\recover2

Them mount c:\tmp\recover2 to a drive letter.

When you look at that mounted drive, you will see the encrypted names, but if you access the files from the mounted drive letter they will be unencrypted on the fly.

If you want to find out what kind of files they are, you need to get something that works like the Unix/Linux "file" command.

It tells you what kind of file a file is by examining the data.  It usually can tell by looking at the first few bytes because most common file formats have a "magic number" there.

I use the file command from cygwin (https://cygwin.com/).  Here are some other suggestions

https://superuser.com/questions/272338/what-is-the-equivalent-to-the-linux-file-command-for-windows

There might be a Windows Explorer shell plugin that has this functionality.

Here is file command run in my normal cppcryptfs mounted test folder.

```
K:\foo>file *
foo.docx: Microsoft Word 2007+
foo.xlsx: Microsoft Excel 2007+
test.pdf: PDF document, version 1.7
test.png: PNG image data, 384 x 518, 8-bit/color RGBA, non-interlaced
```

Here it is run in my "recovered" mounted filder
```
H:\>file *
FOs5CY83vlPz1gmnt3l4sw: Microsoft Excel 2007+
PNPm_jLaytQyYaz5ukn-yA: Microsoft Word 2007+
RqW9eZ6ZexGVcv3wstXxOA: PDF document, version 1.7
oRSkMA7lZfXoJ6XXlMSwKw: PNG image data, 384 x 518, 8-bit/color RGBA, non-interlaced
```

The sort order is different, but the results are the same.

Now, if I, for example, rename PNPm_jLaytQyYaz5ukn-yA to PNPm_jLaytQyYaz5ukn-yA.docx, I can open it in word from the mounted drive letter.

If the file command tells you you have mostly one type of file there you can try renaming them all to the right extension.

e.g. (in the Windows cmd.exe window in c:\tmp\recover2)

```
ren *. *.docx

```

The above command won't rename the gocryptfs.conf. Note the space between the "\*." and the "\*.docx"

Or say you have some word documents and pdfs both, you could then do

```
ren *.docx *.pdf
```

And then see if they open in your pdf viewer (or go by the output of the "file *" command to know which files to try in which application).

**Comment by loungebob** [comment]
"You could also decrypt all the files at once (but not the names) fairly easily."

oh that would have saved me the whole sunday lol. I just did it manually and it all worked out ok. 

thanks for the detailed write-up! much appreciated.

but I dont feel at ease until I know WHY this happens. any idea why there are those encrypted files in the root folder when I dont put them there in the first place? is this a cppcryptfs thing or a google drive file stream thing?
problem is, if they were just duplicates of stuff that has been successfully put in the designated place in folder2, then I'd just delete those in root. I did some checking and sometimes, they are partial uploaded files, sometimes they are just duplicates from files successfully put in folder2 but sometimes they are the original intended uploads that didnt make it to folder2.

I assume it is a google drive thing and has to do with upload limits. but before I talk to google I need more information. especially if it is some kind of overflow from stuff that couldnt be uploaded properly (be it limits or other reasons) why dont they get processed once whatever the problem was has disappeared. those files seem to just stay there and the only reason I discovered them in the first place was by accident.

**Comment by bailey27** [comment]
I'm glad you were able to inspect your files.

I don't know, but I'm 99.9% sure it's Google Drive Stream and not cppcryptfs putting those files there.

I have never used Google Drive Stream because they won't let you use it unless you have a work or school Google account, which I don't have.

I can't imagine any scenario in which cppcryptfs would copy an encrypted file somewhere else on its own.

**Comment by loungebob** [comment]
okidoki. I’ll have a chat with google drive support and see what gives. if there’s some info I’ll post it here.
